File: client/common/protocol.go
package common

import (
	"encoding/binary"
	"errors"
	"fmt"
	"strconv"
)

// Package common provides protocol definitions and message handling for the betting system.
//
// Protocol Overview:
// - All messages start with a message type byte
// - For batch bet messages: agency number, batch number, bets count, then bet data
// - Agency number is a single byte (1 character)
// - Fixed fields (DNI, Birthday) have predefined sizes and no length prefix
// - Variable fields (Name, Surname) have a length byte followed by content
// - Maximum variable field size is 255 bytes
// - BetNumber is a fixed 4-byte field stored as big endian 32-bit integer
//
// Message Types:
// - MessageTypeBet (1): Client sends bet information to server (now supports batches)
// - MessageTypeAck (2): Server acknowledges batch of bets sent (type + agency + batch number)
// - MessageTypeClose (3): Graceful connection shutdown
// - MessageTypeNotification (4): Client notifies completion of bet sending
// - MessageTypeWinnersRequest (5): Client requests winners
// - MessageTypeWinnersNotAvailable (6): Server responds winners not ready
// - MessageTypeWinnersAvailable (7): Server responds with winners
//
// Batch Bet Message Format:
// 1st byte: message type (MessageTypeBet)
// 2nd byte: agency number
// 3rd-6th bytes: batch number (4-byte big endian integer)
// 7th byte: number of bets (n)
// Following bytes: n times (name + surname + DNI + birthday + bet number)
//
// ACK Message Format:
// 1st byte: message type (MessageTypeAck)
// 2nd byte: agency number
// 3rd-6th bytes: batch number (4-byte big endian integer)

// Protocol message types
const (
	MessageTypeBet                 = 1
	MessageTypeAck                 = 2
	MessageTypeClose               = 3
	MessageTypeNotification        = 4 // Client notifies completion of bet sending
	MessageTypeNotificationAck     = 5 // Server acknowledges notification
	MessageTypeWinnersRequest      = 6 // Client requests winners
	MessageTypeWinnersNotAvailable = 7 // Server responds winners not ready
	MessageTypeWinnersAvailable    = 8 // Server responds with winners
)

// Common protocol field sizes
const (
	MessageTypeSize      = 1
	AgencyNumberSize     = 1
	BatchNumberSize      = 4
	BetsCountSize        = 1
	DNISize              = 8
	BirthdaySize         = 10
	BetNumberSize        = 4
	MaxVariableFieldSize = 255
)

// ACK message protocol constants
const (
	// ACK message structure: [type][agency][batch_number]
	AckSize = MessageTypeSize + AgencyNumberSize + BatchNumberSize

	// Byte offsets in ACK message (after message type byte)
	AckAgencyOffset      = MessageTypeSize
	AckBatchNumberOffset = MessageTypeSize + AgencyNumberSize
)

// AckBetMessage represents an acknowledgment message for a batch of bets
type AckBetMessage struct {
	AgencyNumber byte
	BatchNumber  uint32 // Changed to uint32 for 4-byte batch number
}

// Error message constants
const (
	ErrInsufficientData = "insufficient data"
	ErrInvalidBetNumber = "invalid bet number: cannot be empty"
	ErrEmptyBetNumber   = "invalid BetNumber field: cannot be empty"
)

// DeserializeAckBetMessage deserializes ACK bet message according to the protocol:
// 1st byte: message type (MessageTypeAck)
// 2nd byte: agency number
// 3rd-6th bytes: batch number (4-byte big endian integer)
func DeserializeAckBetMessage(data []byte) (*AckBetMessage, error) {
	if len(data) < AckSize {
		return nil, fmt.Errorf("data too short: need at least %d bytes", AckSize)
	}

	// Validate message type
	if err := validateMessageType(data, MessageTypeAck); err != nil {
		return nil, err
	}

	ackMsg := &AckBetMessage{}

	// Extract fields using offset constants
	ackMsg.AgencyNumber = data[AckAgencyOffset]

	// Extract 4-byte batch number (big endian)
	batchNumberBytes := data[AckBatchNumberOffset : AckBatchNumberOffset+BatchNumberSize]
	ackMsg.BatchNumber = binary.BigEndian.Uint32(batchNumberBytes)

	return ackMsg, nil
}

// CreateCloseMessage creates a close message to gracefully shutdown the connection
func CreateCloseMessage() []byte {
	return []byte{MessageTypeClose}
}

// ReceiveAckMessage handles the protocol for receiving ACK messages from the server
func ReceiveAckMessage(connection Connection) (*AckBetMessage, error) {

	data, err := connection.ReceiveExactBytes(AckSize)
	if err != nil {
		return nil, fmt.Errorf("error reading ACK data: %v", err)
	}

	// Deserialize the complete ACK message
	return DeserializeAckBetMessage(data)
}

// SendBetMessage handles the protocol for sending bet messages to the server
func SendBetMessage(connection Connection, message *Message) error {
	return connection.Send(message.Content)
}

// SendCloseMessage handles the protocol for sending close messages to the server
func SendCloseMessage(connection Connection) error {
	closeMessage := CreateCloseMessage()
	return connection.Send(closeMessage)
}

// SendBatchBetMessage handles the protocol for sending batch bet messages to the server
func SendBatchBetMessage(connection Connection, agencyNumber string, batchNumber int, bets []Bet) error {
	batchMessage, err := SerializeBatchBetMessage(agencyNumber, batchNumber, bets)
	if err != nil {
		return fmt.Errorf("error serializing batch bet message: %v", err)
	}
	return connection.Send(batchMessage)
}

// CreateNotificationMessage creates a notification message to inform server that all bets have been sent
func CreateNotificationMessage(agencyNumber string) []byte {
	agencyNum := extractAgencyNumber(agencyNumber)
	return []byte{MessageTypeNotification, agencyNum}
}

// CreateWinnersRequestMessage creates a request message to ask for winners
func CreateWinnersRequestMessage(agencyNumber string) []byte {
	agencyNum := extractAgencyNumber(agencyNumber)
	return []byte{MessageTypeWinnersRequest, agencyNum}
}

// SendNotificationMessage sends a notification to the server that all bets have been sent
func SendNotificationMessage(connection Connection, agencyNumber string) error {
	notificationMessage := CreateNotificationMessage(agencyNumber)
	return connection.Send(notificationMessage)
}

// SendWinnersRequestMessage sends a winners request to the server
func SendWinnersRequestMessage(connection Connection, agencyNumber string) error {
	winnersRequestMessage := CreateWinnersRequestMessage(agencyNumber)
	return connection.Send(winnersRequestMessage)
}

// ReceiveWinnersResponse receives the response to a winners request
func ReceiveWinnersResponse(connection Connection) (bool, []string, error) {
	// First read the message type
	messageTypeData, err := connection.ReceiveExactBytes(1)
	if err != nil {
		return false, nil, fmt.Errorf("error reading message type: %v", err)
	}

	messageType := messageTypeData[0]

	switch messageType {
	case MessageTypeWinnersNotAvailable:
		return false, nil, nil

	case MessageTypeWinnersAvailable:
		// Read number of winners
		numWinnersData, err := connection.ReceiveExactBytes(1)
		if err != nil {
			return false, nil, fmt.Errorf("error reading number of winners: %v", err)
		}
		numWinners := int(numWinnersData[0])

		// Read each winner DNI (8 bytes each)
		winners := make([]string, numWinners)
		for i := 0; i < numWinners; i++ {
			dniData, err := connection.ReceiveExactBytes(DNISize)
			if err != nil {
				return false, nil, fmt.Errorf("error reading winner DNI %d: %v", i, err)
			}
			// Trim whitespace from DNI
			winners[i] = string(dniData)
			winners[i] = winners[i][:len(winners[i])-countTrailingSpaces(winners[i])]
		}

		return true, winners, nil

	default:
		return false, nil, fmt.Errorf("unexpected message type: %d", messageType)
	}
}

// countTrailingSpaces counts trailing spaces in a string
func countTrailingSpaces(s string) int {
	count := 0
	for i := len(s) - 1; i >= 0; i-- {
		if s[i] == ' ' {
			count++
		} else {
			break
		}
	}
	return count
}

// ReceiveNotificationResponse receives the response to a notification message
func ReceiveNotificationResponse(connection Connection) error {
	// Read notification ACK response
	data, err := connection.ReceiveExactBytes(1)
	if err != nil {
		return fmt.Errorf("error reading notification response: %v", err)
	}

	if data[0] != MessageTypeNotificationAck {
		return fmt.Errorf("unexpected response type: %d, expected %d", data[0], MessageTypeNotificationAck)
	}

	return nil
}

// serializeBetNumberField converts a bet number string to 4-byte big endian format
func serializeBetNumberField(result []byte, betNumber string) ([]byte, error) {
	// Convert string to integer
	betInt, err := strconv.ParseUint(betNumber, 10, 32)
	if err != nil {
		return nil, fmt.Errorf("invalid bet number: %s is not a valid integer", betNumber)
	}

	// Convert to 4-byte big endian
	betBytes := make([]byte, 4)
	binary.BigEndian.PutUint32(betBytes, uint32(betInt))
	result = append(result, betBytes...)

	return result, nil
}

// Helper functions for common operations

// serializeVariableField adds a variable length field to the result
// Format: [length_byte][content]
func serializeVariableField(result []byte, content string) []byte {
	contentBytes := []byte(content)
	if len(contentBytes) > MaxVariableFieldSize {
		contentBytes = contentBytes[:MaxVariableFieldSize]
	}
	result = append(result, byte(len(contentBytes)))
	result = append(result, contentBytes...)
	return result
}

// serializeFixedField adds a fixed length field to the result and validates its size
func serializeFixedField(result []byte, content string, expectedSize int, fieldName string) ([]byte, error) {
	contentBytes := []byte(content)
	if len(contentBytes) != expectedSize {
		return nil, fmt.Errorf("invalid %s field: must be %d characters", fieldName, expectedSize)
	}
	result = append(result, contentBytes...)
	return result, nil
}

// validateMessageType checks if the first byte matches the expected message type
func validateMessageType(data []byte, expectedType byte) error {
	if len(data) == 0 {
		return errors.New("empty data")
	}
	if data[0] != expectedType {
		return fmt.Errorf("invalid message type: expected %d, got %d", expectedType, data[0])
	}
	return nil
}

// extractAgencyNumber safely extracts agency number from AgencyInfo ID
func extractAgencyNumber(id string) byte {
	if len(id) > 0 {
		if agencyNum, err := strconv.ParseUint(id, 10, 8); err == nil {
			return byte(agencyNum)
		}
	}
	return 0 // Default fallback
}

// SerializeBatchBetMessage converts a batch of bets to bytes according to the new protocol:
// 1st byte: message type (MessageTypeBet)
// 2nd byte: agency number
// 3rd byte: batch number
// 4th byte: number of bets (n)
// Following bytes: n times (name + surname + DNI + birthday + bet number)
func SerializeBatchBetMessage(agencyNumber string, batchNumber int, bets []Bet) ([]byte, error) {
	var result []byte
	var err error

	// 1st byte: message type
	result = append(result, MessageTypeBet)

	// 2nd byte: agency number
	agencyNum := extractAgencyNumber(agencyNumber)
	result = append(result, agencyNum)

	// 3rd-6th bytes: batch number (4-byte big endian integer)
	if batchNumber < 0 || batchNumber > 0xFFFFFFFF {
		return nil, fmt.Errorf("batch number out of range: %d (must be 0 to %d)", batchNumber, 0xFFFFFFFF)
	}
	batchBytes := make([]byte, 4)
	binary.BigEndian.PutUint32(batchBytes, uint32(batchNumber))
	result = append(result, batchBytes...)

	// 7th byte: number of bets
	numBets := len(bets)
	if numBets > 255 {
		return nil, fmt.Errorf("too many bets in batch: %d (max 255)", numBets)
	}
	result = append(result, byte(numBets))

	// Serialize each bet
	for _, bet := range bets {
		// Serialize variable fields (name and surname)
		result = serializeVariableField(result, bet.Name)
		result = serializeVariableField(result, bet.Surname)

		// Serialize fixed fields (DNI and birthday)
		result, err = serializeFixedField(result, bet.DNI, DNISize, "DNI")
		if err != nil {
			return nil, err
		}

		result, err = serializeFixedField(result, bet.Birthday, BirthdaySize, "Birthday")
		if err != nil {
			return nil, err
		}

		// Serialize BetNumber field (4-byte big endian integer)
		if len(bet.BetNumber) == 0 {
			return nil, fmt.Errorf("invalid BetNumber field: cannot be empty")
		}
		result, err = serializeBetNumberField(result, bet.BetNumber)
		if err != nil {
			return nil, err
		}
	}

	return result, nil
}

File: server/common/protocol.py
"""
Protocol module for the betting system server.

This module handles message serialization, deserialization, and protocol constants
for communication between client and server.

Protocol Overview:
- All messages start with a message type byte
- Fixed fields (DNI, Birthday) have predefined sizes and no length prefix
- Variable fields (Name, Surname) have a length byte followed by content
- BetNumber is a fixed 4-byte field stored as big endian 32-bit integer
- Maximum variable field size is 255 bytes

Message Types:
- MESSAGE_TYPE_BET (1): Client sends bet information to server
- MESSAGE_TYPE_ACK (2): Server acknowledges bet receipt
- MESSAGE_TYPE_CLOSE (3): Graceful connection shutdown
"""

import logging
from .utils import Bet
from .transport import TCPConnection

# Protocol message types (matching the client constants)
MESSAGE_TYPE_BET = 1
MESSAGE_TYPE_ACK = 2
MESSAGE_TYPE_CLOSE = 3

# Common protocol field sizes
MESSAGE_TYPE_SIZE = 1
AGENCY_NUMBER_SIZE = 1
DNI_SIZE = 8
BIRTHDAY_SIZE = 10
BET_NUMBER_SIZE = 4
MAX_VARIABLE_FIELD_SIZE = 255
MAX_BYTE_NUMBER_SIZE = 255

# Error message constants
ERR_INSUFFICIENT_DATA = "insufficient data"
ERR_INVALID_BET_NUMBER = "invalid bet number: cannot be empty"
ERR_EMPTY_BET_NUMBER = "invalid BetNumber field: cannot be empty"

class AckBetMessage:
    """Represents an acknowledgment message for a bet"""

    def __init__(self, agency_number: str, dni: str, bet_number: str):
        self.agency_number = agency_number
        self.dni = dni
        self.bet_number = bet_number

    def serialize_ack_bet_message(self):
        """
        Serializes the ACK bet message according to the protocol:
        1st byte: message type (MESSAGE_TYPE_ACK)
        2nd byte: agency number
        Following 8 bytes: DNI (padded with spaces if shorter)
        Following 4 bytes: bet number (big endian 32-bit integer)
        """
        if not self.bet_number or len(self.bet_number) == 0:
            raise ValueError(ERR_EMPTY_BET_NUMBER)
        
        result = bytearray()
        
        # Add message type as first byte
        result.append(MESSAGE_TYPE_ACK)
        
        # Add agency number as second byte
        try:
            agency_num = int(self.agency_number)
            if 0 <= agency_num <= MAX_BYTE_NUMBER_SIZE:
                result.append(agency_num)
            else:
                result.append(0)  # Default fallback for out of range
        except ValueError:
            result.append(0)  # Default fallback for invalid number
        
        # Add DNI as fixed 8 bytes
        dni_bytes = serialize_fixed_field(self.dni, DNI_SIZE, "DNI")
        result.extend(dni_bytes)
        
        # Add bet number as 4-byte big endian integer
        bet_number_bytes = serialize_bet_number_field(self.bet_number)
        result.extend(bet_number_bytes)
        
        return bytes(result)

def serialize_variable_field(content: str) -> bytes:
    """Serializes a variable length field with length prefix"""
    content_bytes = content.encode('utf-8')
    if len(content_bytes) > MAX_VARIABLE_FIELD_SIZE:
        content_bytes = content_bytes[:MAX_VARIABLE_FIELD_SIZE]
    
    result = bytearray()
    result.append(len(content_bytes))
    result.extend(content_bytes)
    return bytes(result)

def serialize_fixed_field(content: str, expected_size: int, field_name: str) -> bytes:
    """Serializes a fixed length field and validates its size"""
    content_bytes = content.encode('utf-8')
    if len(content_bytes) != expected_size:
        raise ValueError(f"invalid {field_name} field: must be {expected_size} characters")
    return content_bytes

def validate_message_type(data: bytes, expected_type: int) -> None:
    """Validates the message type byte"""
    if len(data) == 0:
        raise ValueError("empty data")
    if data[0] != expected_type:
        raise ValueError(f"invalid message type: expected {expected_type}, got {data[0]}")

def deserialize_bet_message(data: bytes) -> Bet:
    """
    Deserializes bet message according to the protocol:

    1st byte: message type (MESSAGE_TYPE_BET)

    2nd byte: agency number

    Variable length fields (Name, Surname): 1 byte length + string content

    Fixed length fields (DNI=8 chars, Birthday=10 chars): direct content

    BetNumber: 4-byte big endian unsigned integer
    
    Returns a Bet object from utils.py
    """
    if len(data) < 2:
        raise ValueError("data too short: missing message type and agency number")
    
    pos = 0
    
    # Check message type
    validate_message_type(data, MESSAGE_TYPE_BET)
    pos += 1
    
    # Extract agency number (second byte)
    agency_number = data[pos]
    pos += 1
    
    # Deserialize Name field: length byte + content
    if pos >= len(data):
        raise ValueError("data too short: missing length for Name field")
    name_length = data[pos]
    pos += 1
    if pos + name_length > len(data):
        raise ValueError("data too short: missing content for Name field")
    first_name = data[pos:pos + name_length].decode('utf-8')
    pos += name_length
    
    # Deserialize Surname field: length byte + content
    if pos >= len(data):
        raise ValueError("data too short: missing length for Surname field")
    surname_length = data[pos]
    pos += 1
    if pos + surname_length > len(data):
        raise ValueError("data too short: missing content for Surname field")
    last_name = data[pos:pos + surname_length].decode('utf-8')
    pos += surname_length
    
    # Deserialize DNI field: fixed characters, no length byte
    if pos + DNI_SIZE > len(data):
        raise ValueError(f"data too short: missing DNI field ({DNI_SIZE} chars)")
    document = data[pos:pos + DNI_SIZE].decode('utf-8').rstrip()
    pos += DNI_SIZE
    
    # Deserialize Birthday field: fixed characters, no length byte
    if pos + BIRTHDAY_SIZE > len(data):
        raise ValueError(f"data too short: missing Birthday field ({BIRTHDAY_SIZE} chars)")
    birthdate = data[pos:pos + BIRTHDAY_SIZE].decode('utf-8').rstrip()
    pos += BIRTHDAY_SIZE
    
    # Deserialize BetNumber field: fixed 4-byte big endian integer
    if pos + BET_NUMBER_SIZE > len(data):
        raise ValueError(f"data too short: missing BetNumber field ({BET_NUMBER_SIZE} bytes)")
    bet_number_bytes = data[pos:pos + BET_NUMBER_SIZE]
    number = deserialize_bet_number_field(bet_number_bytes)
    pos += BET_NUMBER_SIZE
    
    # Create and return Bet object using the constructor from utils.py
    return Bet(
        agency=agency_number,
        first_name=first_name,
        last_name=last_name,
        document=document,
        birthdate=birthdate,
        number=number
    )

def create_ack_message(bet: Bet) -> AckBetMessage:
    """Creates an ACK message from bet information"""
    if not str(bet.number) or len(str(bet.number)) == 0:
        raise ValueError("cannot create ACK for bet with empty bet number")
    
    return AckBetMessage(
        agency_number=str(bet.agency),
        dni=bet.document,
        bet_number=str(bet.number)
    )

def receive_message_type(connection: TCPConnection) -> int:
    """
    Receives and returns the message type from the connection
    """
    try:
        message_type_data = connection.receive_exact_bytes(1)
        return message_type_data[0]
    except Exception as e:
        logging.error(f"action: receive_message_type | result: fail | error: {e}")
        raise

def receive_bet_message_from_connection(connection: TCPConnection) -> bytes:
    """
    Receives a bet message from connection using protocol knowledge
    Returns the complete message data as bytes
    """
    try:
        # Start by reading the message type (1 byte)
        message_type_data = connection.receive_exact_bytes(1)
        message_type = message_type_data[0]
        
        # Verify it's a bet message
        if message_type != MESSAGE_TYPE_BET:
            raise ValueError(f"expected bet message type {MESSAGE_TYPE_BET}, got {message_type}")
        
        # Read the agency number (1 byte)
        agency_data = connection.receive_exact_bytes(AGENCY_NUMBER_SIZE)
        
        # Read name length and name content
        name_length_data = connection.receive_exact_bytes(1)
        name_length = name_length_data[0]
        name_data = connection.receive_exact_bytes(name_length) if name_length > 0 else b''
        
        # Read surname length and surname content
        surname_length_data = connection.receive_exact_bytes(1)
        surname_length = surname_length_data[0]
        surname_data = connection.receive_exact_bytes(surname_length) if surname_length > 0 else b''
        
        # Read fixed DNI
        dni_data = connection.receive_exact_bytes(DNI_SIZE)
        
        # Read fixed Birthday
        birthday_data = connection.receive_exact_bytes(BIRTHDAY_SIZE)
        
        # Read bet number as fixed 4-byte big endian integer
        bet_data = connection.receive_exact_bytes(BET_NUMBER_SIZE)
        
        # Combine all data according to protocol (including message type)
        full_message = (message_type_data + agency_data + 
                       name_length_data + name_data +
                       surname_length_data + surname_data +
                       dni_data + 
                       birthday_data + 
                       bet_data)
        
        logging.info(f"action: bet_message_received | result: success | total_bytes: {len(full_message)}")
        return full_message
        
    except Exception as e:
        logging.error(f"action: receive_bet_message_from_connection | result: fail | error: {e}")
        raise

def send_ack_message_to_connection(connection: TCPConnection, ack_data: bytes) -> None:
    """
    Sends ACK message to connection
    Handles the low-level sending with proper error handling
    """
    try:
        connection.send(ack_data)
        logging.info(f"action: ack_message_sent | result: success | bytes: {len(ack_data)}")
    except Exception as e:
        logging.error(f"action: send_ack_message_to_connection | result: fail | error: {e}")
        raise

def serialize_bet_number_field(bet_number: str) -> bytes:
    """Converts a bet number string to 4-byte big endian format"""
    try:
        bet_int = int(bet_number)
        if bet_int < 0 or bet_int > 0xFFFFFFFF:
            raise ValueError(f"bet number {bet_number} is out of range for 32-bit integer")
        return bet_int.to_bytes(4, byteorder='big')
    except ValueError as e:
        raise ValueError(f"invalid bet number: {bet_number} is not a valid integer") from e

def deserialize_bet_number_field(bet_bytes: bytes) -> str:
    """Converts 4-byte big endian format to bet number string"""
    if len(bet_bytes) != 4:
        raise ValueError(f"bet number field must be exactly 4 bytes, got {len(bet_bytes)}")
    
    bet_int = int.from_bytes(bet_bytes, byteorder='big')
    return str(bet_int)
